// novo-elements/esm2015/elements/simple-table/cell-header.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Directive, ElementRef, Input, Optional, ViewChild, ViewEncapsulation, } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { CdkColumnDef } from '@angular/cdk/table';
import * as dateFns from 'date-fns';
import { NovoDropdownElement } from '../dropdown/Dropdown';
import { NovoSortFilter } from './sort';
import { NovoLabelService } from '../../services/novo-label-service';
import { NovoActivityTableState } from './state';
import { Helpers } from '../../utils/Helpers';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../services/novo-label-service';
import * as ɵngcc2 from './state';
import * as ɵngcc3 from './sort';
import * as ɵngcc4 from '@angular/cdk/table';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../button/Button';
import * as ɵngcc7 from '../dropdown/Dropdown';
import * as ɵngcc8 from '../date-picker/DatePicker';
import * as ɵngcc9 from '@angular/forms';

const _c0 = ["novo-simple-cell-config", ""];
function NovoSimpleCellHeader_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function NovoSimpleCellHeader_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.sort(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("active", ctx_r0.sortActive);
    ɵngcc0.ɵɵproperty("icon", ctx_r0.icon);
} }
function NovoSimpleCellHeader_novo_dropdown_4_button_5_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 11);
    ɵngcc0.ɵɵlistener("click", function NovoSimpleCellHeader_novo_dropdown_4_button_5_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(2); return ctx_r8.clearFilter(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r4.labels.clear, " ");
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_7_ng_container_1_item_1_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 18);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_7_ng_container_1_item_1_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "item", 16);
    ɵngcc0.ɵɵlistener("click", function NovoSimpleCellHeader_novo_dropdown_4_list_7_ng_container_1_item_1_Template_item_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); const option_r14 = ctx.$implicit; const ctx_r16 = ɵngcc0.ɵɵnextContext(4); return ctx_r16.filterData(option_r14); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, NovoSimpleCellHeader_novo_dropdown_4_list_7_ng_container_1_item_1_i_2_Template, 1, 0, "i", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const option_r14 = ctx.$implicit;
    const ctx_r13 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵclassProp("active", ctx_r13.activeDateFilter === option_r14.label);
    ɵngcc0.ɵɵattribute("data-automation-id", "novo-activity-table-filter-" + option_r14.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", option_r14.label, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r13.activeDateFilter === option_r14.label);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_7_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, NovoSimpleCellHeader_novo_dropdown_4_list_7_ng_container_1_item_1_Template, 3, 5, "item", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r10.config.filterConfig.options);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_7_item_2_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 18);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_7_item_2_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "item", 19);
    ɵngcc0.ɵɵlistener("click", function NovoSimpleCellHeader_novo_dropdown_4_list_7_item_2_Template_item_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(3); return ctx_r19.toggleCustomRange($event, true); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, NovoSimpleCellHeader_novo_dropdown_4_list_7_item_2_i_2_Template, 1, 0, "i", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassProp("active", ctx_r11.labels.customDateRange === ctx_r11.activeDateFilter);
    ɵngcc0.ɵɵproperty("keepOpen", true);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r11.labels.customDateRange, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.labels.customDateRange === ctx_r11.activeDateFilter);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_7_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵelementStart(1, "div", 16);
    ɵngcc0.ɵɵlistener("click", function NovoSimpleCellHeader_novo_dropdown_4_list_7_div_3_Template_div_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(3); return ctx_r21.toggleCustomRange($event, false); });
    ɵngcc0.ɵɵelement(2, "i", 21);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "novo-date-picker", 22);
    ɵngcc0.ɵɵlistener("onSelect", function NovoSimpleCellHeader_novo_dropdown_4_list_7_div_3_Template_novo_date_picker_onSelect_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r23 = ɵngcc0.ɵɵnextContext(3); return ctx_r23.filterData($event); })("ngModelChange", function NovoSimpleCellHeader_novo_dropdown_4_list_7_div_3_Template_novo_date_picker_ngModelChange_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r24 = ɵngcc0.ɵɵnextContext(3); return ctx_r24.filter = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r12.labels.backToPresetFilters);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r12.filter);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "list");
    ɵngcc0.ɵɵtemplate(1, NovoSimpleCellHeader_novo_dropdown_4_list_7_ng_container_1_Template, 2, 1, "ng-container", 12);
    ɵngcc0.ɵɵtemplate(2, NovoSimpleCellHeader_novo_dropdown_4_list_7_item_2_Template, 3, 5, "item", 13);
    ɵngcc0.ɵɵtemplate(3, NovoSimpleCellHeader_novo_dropdown_4_list_7_div_3_Template, 5, 2, "div", 14);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r5.showCustomRange);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.config.filterConfig.allowCustomRange && !ctx_r5.showCustomRange);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.showCustomRange);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_8_item_1_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 18);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_8_item_1_Template(rf, ctx) { if (rf & 1) {
    const _r29 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "item", 16);
    ɵngcc0.ɵɵlistener("click", function NovoSimpleCellHeader_novo_dropdown_4_list_8_item_1_Template_item_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r29); const option_r26 = ctx.$implicit; const ctx_r28 = ɵngcc0.ɵɵnextContext(3); return ctx_r28.filterData(option_r26); });
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, NovoSimpleCellHeader_novo_dropdown_4_list_8_item_1_i_3_Template, 1, 0, "i", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const option_r26 = ctx.$implicit;
    const ctx_r25 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassProp("active", ctx_r25.filter === option_r26);
    ɵngcc0.ɵɵattribute("data-automation-id", "novo-activity-table-filter-" + ((option_r26 == null ? null : option_r26.label) || option_r26));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate((option_r26 == null ? null : option_r26.label) || option_r26);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", option_r26.hasOwnProperty("value") ? ctx_r25.filter === option_r26.value : ctx_r25.filter === option_r26);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "list");
    ɵngcc0.ɵɵtemplate(1, NovoSimpleCellHeader_novo_dropdown_4_list_8_item_1_Template, 4, 5, "item", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r6.config.filterConfig.options);
} }
function NovoSimpleCellHeader_novo_dropdown_4_list_9_Template(rf, ctx) { if (rf & 1) {
    const _r31 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "list");
    ɵngcc0.ɵɵelementStart(1, "item", 23);
    ɵngcc0.ɵɵelementStart(2, "input", 24);
    ɵngcc0.ɵɵlistener("ngModelChange", function NovoSimpleCellHeader_novo_dropdown_4_list_9_Template_input_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r31); const ctx_r30 = ɵngcc0.ɵɵnextContext(2); return ctx_r30.filter = $event; })("ngModelChange", function NovoSimpleCellHeader_novo_dropdown_4_list_9_Template_input_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r31); const ctx_r32 = ɵngcc0.ɵɵnextContext(2); return ctx_r32.filterData($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r7.filter);
} }
function NovoSimpleCellHeader_novo_dropdown_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "novo-dropdown", 4);
    ɵngcc0.ɵɵelement(1, "button", 5);
    ɵngcc0.ɵɵelementStart(2, "div", 6);
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, NovoSimpleCellHeader_novo_dropdown_4_button_5_Template, 2, 1, "button", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerStart(6, 8);
    ɵngcc0.ɵɵtemplate(7, NovoSimpleCellHeader_novo_dropdown_4_list_7_Template, 4, 3, "list", 9);
    ɵngcc0.ɵɵtemplate(8, NovoSimpleCellHeader_novo_dropdown_4_list_8_Template, 2, 1, "list", 9);
    ɵngcc0.ɵɵtemplate(9, NovoSimpleCellHeader_novo_dropdown_4_list_9_Template, 3, 1, "list", 10);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("active", ctx_r1.filterActive);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.labels.filters);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.filter);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitch", ctx_r1.config.filterConfig.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "date");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "select");
} }
const _c1 = ["*"];
export class NovoSimpleFilterFocus {
    constructor(element) {
        this.element = element;
    }
    ngAfterViewInit() {
        this.element.nativeElement.focus();
    }
}
NovoSimpleFilterFocus.ɵfac = function NovoSimpleFilterFocus_Factory(t) { return new (t || NovoSimpleFilterFocus)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NovoSimpleFilterFocus.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NovoSimpleFilterFocus, selectors: [["", "novoSimpleFilterFocus", ""]] });
NovoSimpleFilterFocus.ctorParameters = () => [
    { type: ElementRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoSimpleFilterFocus, [{
        type: Directive,
        args: [{
                selector: '[novoSimpleFilterFocus]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
export class NovoSimpleCellHeader {
    constructor(changeDetectorRef, labels, state, _sort, _cdkColumnDef) {
        this.changeDetectorRef = changeDetectorRef;
        this.labels = labels;
        this.state = state;
        this._sort = _sort;
        this._cdkColumnDef = _cdkColumnDef;
        this.icon = 'sortable';
        this.filterActive = false;
        this.sortActive = false;
        this.showCustomRange = false;
        this._rerenderSubscription = state.updates.subscribe((change) => {
            if (change.sort && change.sort.id === this.id) {
                this.icon = `sort-${change.sort.value}`;
                this.sortActive = true;
            }
            else {
                this.icon = 'sortable';
                this.sortActive = false;
            }
            if (change.filter && change.filter.id === this.id) {
                this.filterActive = true;
                this.filter = change.filter.value;
            }
            else {
                this.filterActive = false;
                this.filter = undefined;
            }
            changeDetectorRef.markForCheck();
        });
    }
    get config() {
        return this._config;
    }
    set config(v) {
        if (!v) {
            this._config = {
                sortable: false,
                filterable: false,
                filterConfig: {
                    type: 'text',
                },
            };
        }
        else {
            this._config = {
                sortable: coerceBooleanProperty(v.sortable),
                filterable: coerceBooleanProperty(v.filterable),
                transforms: v.transforms || {},
                filterConfig: v.filterConfig || {
                    type: 'text',
                },
            };
            if (this._config.filterConfig.type === 'date' && !this._config.filterConfig.options) {
                this._config.filterConfig.options = this.getDefaultDateFilterOptions();
            }
        }
    }
    ngOnInit() {
        if (this._cdkColumnDef) {
            this.id = this._cdkColumnDef.name;
        }
        if (this.defaultSort && this.id === this.defaultSort.id) {
            this.icon = `sort-${this.defaultSort.value}`;
            this.sortActive = true;
            this.changeDetectorRef.markForCheck();
        }
    }
    ngOnDestroy() {
        this._rerenderSubscription.unsubscribe();
    }
    sort() {
        if (this.changeTimeout) {
            clearTimeout(this.changeTimeout);
        }
        this.changeTimeout = setTimeout(() => {
            this.direction = this.getNextSortDirection(this.direction);
            this._sort.sort(this.id, this.direction, this._config.transforms.sort);
            this.changeDetectorRef.markForCheck();
        }, 300);
    }
    toggleCustomRange(event, value) {
        Helpers.swallowEvent(event);
        this.showCustomRange = value;
        this.changeDetectorRef.markForCheck();
        this.dropdown.openPanel(); // Ensures that the panel correctly updates to the dynamic size of the dropdown
    }
    filterData(filter) {
        let actualFilter = filter;
        if (this.config.filterConfig.type === 'date' && filter) {
            this.activeDateFilter = filter.label || this.labels.customDateRange;
            if (filter.startDate && filter.endDate) {
                actualFilter = {
                    min: dateFns.startOfDay(filter.startDate.date),
                    max: dateFns.startOfDay(dateFns.addDays(dateFns.startOfDay(filter.endDate.date), 1)),
                };
            }
            else {
                actualFilter = {
                    min: filter.min ? dateFns.addDays(dateFns.startOfToday(), filter.min) : dateFns.startOfToday(),
                    max: filter.max ? dateFns.addDays(dateFns.startOfTomorrow(), filter.max) : dateFns.startOfTomorrow(),
                };
            }
        }
        if (actualFilter && actualFilter.hasOwnProperty('value')) {
            actualFilter = filter.value;
        }
        if (this.changeTimeout) {
            clearTimeout(this.changeTimeout);
        }
        this.changeTimeout = setTimeout(() => {
            if (actualFilter === '') {
                actualFilter = undefined;
            }
            this._sort.filter(this.id, actualFilter, this.config.transforms.filter);
            this.changeDetectorRef.markForCheck();
        }, 300);
    }
    clearFilter() {
        this.filter = undefined;
        this.activeDateFilter = undefined;
        this.filterData();
    }
    getNextSortDirection(direction) {
        if (!direction) {
            return 'asc';
        }
        if (direction === 'asc') {
            return 'desc';
        }
        return 'asc';
    }
    getDefaultDateFilterOptions() {
        const opts = [
            { label: this.labels.past1Day, min: -1, max: 0 },
            { label: this.labels.past7Days, min: -7, max: 0 },
            { label: this.labels.past30Days, min: -30, max: 0 },
            { label: this.labels.past90Days, min: -90, max: 0 },
            { label: this.labels.past1Year, min: -366, max: 0 },
            { label: this.labels.next1Day, min: 0, max: 1 },
            { label: this.labels.next7Days, min: 0, max: 7 },
            { label: this.labels.next30Days, min: 0, max: 30 },
            { label: this.labels.next90Days, min: 0, max: 90 },
            { label: this.labels.next1Year, min: 0, max: 366 },
        ];
        return opts;
    }
}
NovoSimpleCellHeader.ɵfac = function NovoSimpleCellHeader_Factory(t) { return new (t || NovoSimpleCellHeader)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NovoLabelService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.NovoActivityTableState), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.NovoSortFilter, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.CdkColumnDef, 8)); };
NovoSimpleCellHeader.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NovoSimpleCellHeader, selectors: [["", "novo-simple-cell-config", ""]], viewQuery: function NovoSimpleCellHeader_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(NovoDropdownElement, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropdown = _t.first);
    } }, inputs: { config: ["novo-simple-cell-config", "config"], defaultSort: "defaultSort" }, attrs: _c0, ngContentSelectors: _c1, decls: 5, vars: 4, consts: [["data-automation-id", "novo-activity-table-label", 3, "click"], ["theme", "icon", "data-automation-id", "novo-activity-table-sort", 3, "icon", "active", "click", 4, "ngIf"], ["side", "right", "parentScrollSelector", ".novo-simple-table", "containerClass", "simple-table-dropdown", "data-automation-id", "novo-activity-table-filter", 4, "ngIf"], ["theme", "icon", "data-automation-id", "novo-activity-table-sort", 3, "icon", "click"], ["side", "right", "parentScrollSelector", ".novo-simple-table", "containerClass", "simple-table-dropdown", "data-automation-id", "novo-activity-table-filter"], ["type", "button", "theme", "icon", "icon", "filter"], [1, "header"], ["theme", "dialogue", "color", "negative", "icon", "times", "data-automation-id", "novo-activity-table-filter-clear", 3, "click", 4, "ngIf"], [3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngSwitchDefault"], ["theme", "dialogue", "color", "negative", "icon", "times", "data-automation-id", "novo-activity-table-filter-clear", 3, "click"], [4, "ngIf"], [3, "active", "keepOpen", "click", 4, "ngIf"], ["class", "calendar-container", 4, "ngIf"], [3, "active", "click", 4, "ngFor", "ngForOf"], [3, "click"], ["class", "bhi-check", 4, "ngIf"], [1, "bhi-check"], [3, "keepOpen", "click"], [1, "calendar-container"], [1, "bhi-previous"], ["range", "true", 3, "ngModel", "onSelect", "ngModelChange"], ["keepOpen", "true", 1, "filter-search"], ["type", "text", "novoSimpleFilterFocus", "", "data-automation-id", "novo-activity-table-filter-input", 3, "ngModel", "ngModelChange"]], template: function NovoSimpleCellHeader_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "label", 0);
        ɵngcc0.ɵɵlistener("click", function NovoSimpleCellHeader_Template_label_click_0_listener() { return ctx.sort(); });
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(2, "div");
        ɵngcc0.ɵɵtemplate(3, NovoSimpleCellHeader_button_3_Template, 1, 3, "button", 1);
        ɵngcc0.ɵɵtemplate(4, NovoSimpleCellHeader_novo_dropdown_4_Template, 10, 7, "novo-dropdown", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("sort-disabled", !ctx.config.sortable);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.config.sortable);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.config.filterable);
    } }, directives: [ɵngcc5.NgIf, ɵngcc6.NovoButtonElement, ɵngcc7.NovoDropdownElement, ɵngcc5.NgSwitch, ɵngcc5.NgSwitchCase, ɵngcc5.NgSwitchDefault, ɵngcc7.NovoDropdownListElement, ɵngcc5.NgForOf, ɵngcc7.NovoItemElement, ɵngcc8.NovoDatePickerElement, ɵngcc9.NgControlStatus, ɵngcc9.NgModel, ɵngcc9.DefaultValueAccessor, NovoSimpleFilterFocus], encapsulation: 2, changeDetection: 0 });
NovoSimpleCellHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: NovoLabelService },
    { type: NovoActivityTableState },
    { type: NovoSortFilter, decorators: [{ type: Optional }] },
    { type: CdkColumnDef, decorators: [{ type: Optional }] }
];
NovoSimpleCellHeader.propDecorators = {
    dropdown: [{ type: ViewChild, args: [NovoDropdownElement,] }],
    defaultSort: [{ type: Input }],
    config: [{ type: Input, args: ['novo-simple-cell-config',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoSimpleCellHeader, [{
        type: Component,
        args: [{
                selector: '[novo-simple-cell-config]',
                template: `
    <label (click)="sort()" data-automation-id="novo-activity-table-label" [class.sort-disabled]="!config.sortable">
      <ng-content></ng-content>
    </label>
    <div>
      <button
        *ngIf="config.sortable"
        theme="icon"
        [icon]="icon"
        (click)="sort()"
        [class.active]="sortActive"
        data-automation-id="novo-activity-table-sort"
      ></button>
      <novo-dropdown
        *ngIf="config.filterable"
        side="right"
        parentScrollSelector=".novo-simple-table"
        containerClass="simple-table-dropdown"
        data-automation-id="novo-activity-table-filter"
      >
        <button type="button" theme="icon" icon="filter" [class.active]="filterActive"></button>
        <div class="header">
          <span>{{ labels.filters }}</span>
          <button
            theme="dialogue"
            color="negative"
            icon="times"
            (click)="clearFilter()"
            *ngIf="filter"
            data-automation-id="novo-activity-table-filter-clear"
          >
            {{ labels.clear }}
          </button>
        </div>
        <ng-container [ngSwitch]="config.filterConfig.type">
          <list *ngSwitchCase="'date'">
            <ng-container *ngIf="!showCustomRange">
              <item
                [class.active]="activeDateFilter === option.label"
                *ngFor="let option of config.filterConfig.options"
                (click)="filterData(option)"
                [attr.data-automation-id]="'novo-activity-table-filter-' + option.label"
              >
                {{ option.label }} <i class="bhi-check" *ngIf="activeDateFilter === option.label"></i>
              </item>
            </ng-container>
            <item
              [class.active]="labels.customDateRange === activeDateFilter"
              (click)="toggleCustomRange($event, true)"
              *ngIf="config.filterConfig.allowCustomRange && !showCustomRange"
              [keepOpen]="true"
            >
              {{ labels.customDateRange }} <i class="bhi-check" *ngIf="labels.customDateRange === activeDateFilter"></i>
            </item>
            <div class="calendar-container" *ngIf="showCustomRange">
              <div (click)="toggleCustomRange($event, false)"><i class="bhi-previous"></i>{{ labels.backToPresetFilters }}</div>
              <novo-date-picker (onSelect)="filterData($event)" [(ngModel)]="filter" range="true"></novo-date-picker>
            </div>
          </list>
          <list *ngSwitchCase="'select'">
            <item
              [class.active]="filter === option"
              *ngFor="let option of config.filterConfig.options"
              (click)="filterData(option)"
              [attr.data-automation-id]="'novo-activity-table-filter-' + (option?.label || option)"
            >
              <span>{{ option?.label || option }}</span>
              <i class="bhi-check" *ngIf="option.hasOwnProperty('value') ? filter === option.value : filter === option"></i>
            </item>
          </list>
          <list *ngSwitchDefault>
            <item class="filter-search" keepOpen="true">
              <input
                type="text"
                [(ngModel)]="filter"
                (ngModelChange)="filterData($event)"
                novoSimpleFilterFocus
                data-automation-id="novo-activity-table-filter-input"
              />
            </item>
          </list>
        </ng-container>
      </novo-dropdown>
    </div>
  `,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.NovoLabelService }, { type: ɵngcc2.NovoActivityTableState }, { type: ɵngcc3.NovoSortFilter, decorators: [{
                type: Optional
            }] }, { type: ɵngcc4.CdkColumnDef, decorators: [{
                type: Optional
            }] }]; }, { config: [{
            type: Input,
            args: ['novo-simple-cell-config']
        }], dropdown: [{
            type: ViewChild,
            args: [NovoDropdownElement]
        }], defaultSort: [{
            type: Input
        }] }); })();
