// novo-elements/esm2015/elements/data-table/data-table.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { CdkTableModule } from '@angular/cdk/table';
import { NovoButtonModule } from '../button/Button.module';
import { NovoDropdownModule } from '../dropdown/Dropdown.module';
import { NovoFormExtrasModule } from '../form/extras/FormExtras.module';
import { NovoLoadingModule } from '../loading/Loading.module';
import { NovoTilesModule } from '../tiles/Tiles.module';
import { NovoSearchBoxModule } from '../search/SearchBox.module';
import { NovoDatePickerModule } from '../date-picker/DatePicker.module';
import { NovoCommonModule } from '../common/common.module';
import { NovoSelectModule } from '../select/Select.module';
import { NovoTooltipModule } from '../tooltip/Tooltip.module';
import { NovoDataTable } from './data-table.component';
import { NovoDataTableCell } from './cells/data-table-cell.component';
import { NovoDataTableCheckboxCell } from './cells/data-table-checkbox-cell.component';
import { NovoDataTableExpandCell } from './cells/data-table-expand-cell.component';
import { NovoDataTableHeaderRow } from './rows/data-table-header-row.component';
import { NovoDataTableRow } from './rows/data-table-row.component';
import { NovoDataTableCellHeader } from './cell-headers/data-table-header-cell.component';
import { NovoDataTableExpandHeaderCell } from './cell-headers/data-table-expand-header-cell.component';
import { NovoDataTableCheckboxHeaderCell } from './cell-headers/data-table-checkbox-header-cell.component';
import { NovoDataTableHeaderCell } from './cell-headers/data-table-header-cell.directive';
import { NovoDataTableSortFilter } from './sort-filter/sort-filter.directive';
import { NovoDataTablePagination } from './pagination/data-table-pagination.component';
import { DataTableState } from './state/data-table-state.service';
import { DataTableInterpolatePipe, DateTableDateRendererPipe, DateTableCurrencyRendererPipe, DateTableDateTimeRendererPipe, DateTableNumberRendererPipe, DateTableTimeRendererPipe, DataTableBigDecimalRendererPipe, } from './data-table.pipes';
import { NovoDataTableExpandDirective } from './data-table-expand.directive';
import { NovoDataTableClearButton } from './data-table-clear-button.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../date-picker/DatePicker';
import * as ɵngcc2 from '../date-picker/DatePickerInput';
import * as ɵngcc3 from '@angular/cdk/table';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from '../button/Button';
import * as ɵngcc7 from '../dropdown/Dropdown';
import * as ɵngcc8 from '../form/extras/address/Address';
import * as ɵngcc9 from '../form/extras/checkbox/Checkbox';
import * as ɵngcc10 from '../form/extras/checkbox/CheckList';
import * as ɵngcc11 from '../form/extras/file/FileInput';
import * as ɵngcc12 from '../loading/Loading';
import * as ɵngcc13 from '../tiles/Tiles';
import * as ɵngcc14 from '../search/SearchBox';
import * as ɵngcc15 from '../common/novo-template/novo-template.directive';
import * as ɵngcc16 from '../select/Select';
import * as ɵngcc17 from '../tooltip/Tooltip.directive';
export class NovoDataTableModule {
}
NovoDataTableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoDataTableModule });
NovoDataTableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoDataTableModule_Factory(t) { return new (t || NovoDataTableModule)(); }, providers: [DataTableState], imports: [[
            NovoDatePickerModule,
            CdkTableModule,
            CommonModule,
            FormsModule,
            NovoButtonModule,
            NovoDropdownModule,
            NovoFormExtrasModule,
            NovoLoadingModule,
            NovoTilesModule,
            NovoSearchBoxModule,
            NovoCommonModule,
            NovoSelectModule,
            NovoTooltipModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoDataTableModule, { declarations: function () { return [DataTableInterpolatePipe,
        DateTableDateRendererPipe,
        DateTableCurrencyRendererPipe,
        DateTableDateTimeRendererPipe,
        DateTableNumberRendererPipe,
        DateTableTimeRendererPipe,
        DataTableBigDecimalRendererPipe,
        NovoDataTableCellHeader,
        NovoDataTableSortFilter,
        NovoDataTableHeaderCell,
        NovoDataTableCell,
        NovoDataTableHeaderRow,
        NovoDataTableRow,
        NovoDataTablePagination,
        NovoDataTableCheckboxCell,
        NovoDataTableCheckboxHeaderCell,
        NovoDataTableExpandCell,
        NovoDataTableExpandHeaderCell,
        NovoDataTable,
        NovoDataTableExpandDirective,
        NovoDataTableClearButton]; }, imports: function () { return [NovoDatePickerModule,
        CdkTableModule,
        CommonModule,
        FormsModule,
        NovoButtonModule,
        NovoDropdownModule,
        NovoFormExtrasModule,
        NovoLoadingModule,
        NovoTilesModule,
        NovoSearchBoxModule,
        NovoCommonModule,
        NovoSelectModule,
        NovoTooltipModule]; }, exports: function () { return [NovoDataTable,
        DataTableInterpolatePipe,
        DateTableDateRendererPipe,
        DateTableCurrencyRendererPipe,
        DateTableDateTimeRendererPipe,
        DateTableNumberRendererPipe,
        DateTableTimeRendererPipe,
        DataTableBigDecimalRendererPipe,
        NovoDataTableClearButton]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoDataTableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    NovoDatePickerModule,
                    CdkTableModule,
                    CommonModule,
                    FormsModule,
                    NovoButtonModule,
                    NovoDropdownModule,
                    NovoFormExtrasModule,
                    NovoLoadingModule,
                    NovoTilesModule,
                    NovoSearchBoxModule,
                    NovoCommonModule,
                    NovoSelectModule,
                    NovoTooltipModule,
                ],
                declarations: [
                    DataTableInterpolatePipe,
                    DateTableDateRendererPipe,
                    DateTableCurrencyRendererPipe,
                    DateTableDateTimeRendererPipe,
                    DateTableNumberRendererPipe,
                    DateTableTimeRendererPipe,
                    DataTableBigDecimalRendererPipe,
                    NovoDataTableCellHeader,
                    NovoDataTableSortFilter,
                    NovoDataTableHeaderCell,
                    NovoDataTableCell,
                    NovoDataTableHeaderRow,
                    NovoDataTableRow,
                    NovoDataTablePagination,
                    NovoDataTableCheckboxCell,
                    NovoDataTableCheckboxHeaderCell,
                    NovoDataTableExpandCell,
                    NovoDataTableExpandHeaderCell,
                    NovoDataTable,
                    NovoDataTableExpandDirective,
                    NovoDataTableClearButton,
                ],
                providers: [DataTableState],
                exports: [
                    NovoDataTable,
                    DataTableInterpolatePipe,
                    DateTableDateRendererPipe,
                    DateTableCurrencyRendererPipe,
                    DateTableDateTimeRendererPipe,
                    DateTableNumberRendererPipe,
                    DateTableTimeRendererPipe,
                    DataTableBigDecimalRendererPipe,
                    NovoDataTableClearButton,
                ]
            }]
    }], null, null); })();
ɵngcc0.ɵɵsetComponentScope(NovoDataTable, [ɵngcc1.NovoDatePickerElement, ɵngcc2.NovoDatePickerInputElement, ɵngcc3.CdkTable, ɵngcc3.CdkRowDef, ɵngcc3.CdkCellDef, ɵngcc3.CdkCellOutlet, ɵngcc3.CdkHeaderCellDef, ɵngcc3.CdkFooterCellDef, ɵngcc3.CdkColumnDef, ɵngcc3.CdkCell, ɵngcc3.CdkRow, ɵngcc3.CdkHeaderCell, ɵngcc3.CdkFooterCell, ɵngcc3.CdkHeaderRow, ɵngcc3.CdkHeaderRowDef, ɵngcc3.CdkFooterRow, ɵngcc3.CdkFooterRowDef, ɵngcc3.DataRowOutlet, ɵngcc3.HeaderRowOutlet, ɵngcc3.FooterRowOutlet, ɵngcc3.CdkTextColumn, ɵngcc3.CdkNoDataRow, ɵngcc3.NoDataRowOutlet, ɵngcc4.NgClass, ɵngcc4.NgComponentOutlet, ɵngcc4.NgForOf, ɵngcc4.NgIf, ɵngcc4.NgTemplateOutlet, ɵngcc4.NgStyle, ɵngcc4.NgSwitch, ɵngcc4.NgSwitchCase, ɵngcc4.NgSwitchDefault, ɵngcc4.NgPlural, ɵngcc4.NgPluralCase, ɵngcc5.ɵangular_packages_forms_forms_y, ɵngcc5.NgSelectOption, ɵngcc5.ɵangular_packages_forms_forms_x, ɵngcc5.DefaultValueAccessor, ɵngcc5.NumberValueAccessor, ɵngcc5.RangeValueAccessor, ɵngcc5.CheckboxControlValueAccessor, ɵngcc5.SelectControlValueAccessor, ɵngcc5.SelectMultipleControlValueAccessor, ɵngcc5.RadioControlValueAccessor, ɵngcc5.NgControlStatus, ɵngcc5.NgControlStatusGroup, ɵngcc5.RequiredValidator, ɵngcc5.MinLengthValidator, ɵngcc5.MaxLengthValidator, ɵngcc5.PatternValidator, ɵngcc5.CheckboxRequiredValidator, ɵngcc5.EmailValidator, ɵngcc5.NgModel, ɵngcc5.NgModelGroup, ɵngcc5.NgForm, ɵngcc6.NovoButtonElement, ɵngcc7.NovoDropdownElement, ɵngcc7.NovoItemElement, ɵngcc7.NovoDropdownListElement, ɵngcc7.NovoDropDownItemHeaderElement, ɵngcc8.NovoAddressElement, ɵngcc9.NovoCheckboxElement, ɵngcc10.NovoCheckListElement, ɵngcc11.NovoFileInputElement, ɵngcc12.NovoLoadingElement, ɵngcc12.NovoSpinnerElement, ɵngcc12.NovoIsLoadingDirective, ɵngcc12.NovoLoadedDirective, ɵngcc12.NovoSkeletonDirective, ɵngcc13.NovoTilesElement, ɵngcc14.NovoSearchBoxElement, ɵngcc15.NovoTemplate, ɵngcc16.NovoSelectElement, ɵngcc17.TooltipDirective, NovoDataTableCellHeader,
    NovoDataTableSortFilter,
    NovoDataTableHeaderCell,
    NovoDataTableCell,
    NovoDataTableHeaderRow,
    NovoDataTableRow,
    NovoDataTablePagination,
    NovoDataTableCheckboxCell,
    NovoDataTableCheckboxHeaderCell,
    NovoDataTableExpandCell,
    NovoDataTableExpandHeaderCell,
    NovoDataTable,
    NovoDataTableExpandDirective,
    NovoDataTableClearButton], [ɵngcc4.AsyncPipe, ɵngcc4.UpperCasePipe, ɵngcc4.LowerCasePipe, ɵngcc4.JsonPipe, ɵngcc4.SlicePipe, ɵngcc4.DecimalPipe, ɵngcc4.PercentPipe, ɵngcc4.TitleCasePipe, ɵngcc4.CurrencyPipe, ɵngcc4.DatePipe, ɵngcc4.I18nPluralPipe, ɵngcc4.I18nSelectPipe, ɵngcc4.KeyValuePipe, DataTableInterpolatePipe,
    DateTableDateRendererPipe,
    DateTableCurrencyRendererPipe,
    DateTableDateTimeRendererPipe,
    DateTableNumberRendererPipe,
    DateTableTimeRendererPipe,
    DataTableBigDecimalRendererPipe]);
