// novo-elements/esm2015/elements/tip-well/TipWell.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoButtonModule } from './../button/Button.module';
import { NovoTipWellElement } from './TipWell';
import * as ɵngcc0 from '@angular/core';
export class NovoTipWellModule {
}
NovoTipWellModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoTipWellModule });
NovoTipWellModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoTipWellModule_Factory(t) { return new (t || NovoTipWellModule)(); }, imports: [[CommonModule, NovoButtonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoTipWellModule, { declarations: function () { return [NovoTipWellElement]; }, imports: function () { return [CommonModule, NovoButtonModule]; }, exports: function () { return [NovoTipWellElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoTipWellModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NovoButtonModule],
                declarations: [NovoTipWellElement],
                exports: [NovoTipWellElement]
            }]
    }], null, null); })();
