// novo-elements/esm2015/elements/tooltip/Tooltip.component.js
// NG2
import { Component } from '@angular/core';
import { trigger, state, style, animate, transition } from '@angular/animations';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = function (a0, a1, a2, a3, a4) { return [a0, a1, a2, a3, a4]; };
function NovoTooltip_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 2);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@state", ctx_r0.noAnimate ? "no-animation" : "visible")("ngClass", ɵngcc0.ɵɵpureFunction5(3, _c0, ctx_r0.tooltipType, ctx_r0.rounded ? "rounded" : "", ctx_r0.size ? ctx_r0.size : "", ctx_r0.preline ? "preline" : "", ctx_r0.position))("innerHTML", ctx_r0.message, ɵngcc0.ɵɵsanitizeHtml);
} }
function NovoTooltip_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@state", ctx_r1.noAnimate ? "no-animation" : "visible")("ngClass", ɵngcc0.ɵɵpureFunction5(3, _c0, ctx_r1.tooltipType, ctx_r1.rounded ? "rounded" : "", ctx_r1.size ? ctx_r1.size : "", ctx_r1.preline ? "preline" : "", ctx_r1.position));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.message);
} }
export class NovoTooltip {
}
NovoTooltip.ɵfac = function NovoTooltip_Factory(t) { return new (t || NovoTooltip)(); };
NovoTooltip.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NovoTooltip, selectors: [["novo-tooltip"]], decls: 2, vars: 2, consts: [[3, "ngClass", "innerHTML", 4, "ngIf"], [3, "ngClass", 4, "ngIf"], [3, "ngClass", "innerHTML"], [3, "ngClass"]], template: function NovoTooltip_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, NovoTooltip_div_0_Template, 1, 9, "div", 0);
        ɵngcc0.ɵɵtemplate(1, NovoTooltip_div_1_Template, 2, 9, "div", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isHTML);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isHTML);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass], encapsulation: 2, data: { animation: [
            trigger('state', [
                state('initial, void, hidden', style({ opacity: '0' })),
                state('visible', style({ opacity: '1' })),
                transition('* => visible', [
                    style({
                        opacity: 0,
                        visibility: 'visible'
                    }),
                    animate('0.3s ease-in'),
                ]),
                transition('* => hidden', [
                    style({
                        opacity: 1,
                        visibility: 'hidden'
                    }),
                    animate('0.3s ease-in'),
                ]),
            ]),
        ] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoTooltip, [{
        type: Component,
        args: [{
                selector: 'novo-tooltip',
                template: `
    <div *ngIf="this.isHTML" [@state]="noAnimate ? 'no-animation' : 'visible'"
         [ngClass]="[tooltipType, this.rounded ? 'rounded' : '', size ? size : '', this.preline? 'preline' : '', position]"
         [innerHTML]="message"></div>
    <div *ngIf="!this.isHTML" [@state]="noAnimate ? 'no-animation' : 'visible'"
         [ngClass]="[tooltipType, this.rounded ? 'rounded' : '', size ? size : '', this.preline? 'preline' : '', position]">{{message}}</div>`,
                animations: [
                    trigger('state', [
                        state('initial, void, hidden', style({ opacity: '0' })),
                        state('visible', style({ opacity: '1' })),
                        transition('* => visible', [
                            style({
                                opacity: 0,
                                visibility: 'visible'
                            }),
                            animate('0.3s ease-in'),
                        ]),
                        transition('* => hidden', [
                            style({
                                opacity: 1,
                                visibility: 'hidden'
                            }),
                            animate('0.3s ease-in'),
                        ]),
                    ]),
                ]
            }]
    }], null, null); })();
