// novo-elements/esm2015/elements/tooltip/Tooltip.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { TooltipDirective } from './Tooltip.directive';
import { NovoTooltip } from './Tooltip.component';
import * as ɵngcc0 from '@angular/core';
export class NovoTooltipModule {
}
NovoTooltipModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoTooltipModule });
NovoTooltipModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoTooltipModule_Factory(t) { return new (t || NovoTooltipModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoTooltipModule, { declarations: function () { return [TooltipDirective, NovoTooltip]; }, imports: function () { return [CommonModule]; }, exports: function () { return [TooltipDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoTooltipModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [TooltipDirective, NovoTooltip],
                exports: [TooltipDirective]
            }]
    }], null, null); })();
