// novo-elements/esm2015/elements/select/Select.module.js
// NG
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { A11yModule } from '@angular/cdk/a11y';
// App
import { NovoOverlayModule } from '../overlay/Overlay.module';
import { NovoTooltipModule } from '../tooltip/Tooltip.module';
import { NovoSelectElement } from './Select';
import * as ɵngcc0 from '@angular/core';
export class NovoSelectModule {
}
NovoSelectModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoSelectModule });
NovoSelectModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoSelectModule_Factory(t) { return new (t || NovoSelectModule)(); }, imports: [[CommonModule, FormsModule, A11yModule, NovoOverlayModule, NovoTooltipModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoSelectModule, { declarations: function () { return [NovoSelectElement]; }, imports: function () { return [CommonModule, FormsModule, A11yModule, NovoOverlayModule, NovoTooltipModule]; }, exports: function () { return [NovoSelectElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoSelectModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, A11yModule, NovoOverlayModule, NovoTooltipModule],
                declarations: [NovoSelectElement],
                exports: [NovoSelectElement]
            }]
    }], null, null); })();
