// novo-elements/esm2015/pipes/Pipes.module.js
// NG2
import { NgModule } from '@angular/core';
// APP
import { PluralPipe } from './plural/Plural';
import { DecodeURIPipe } from './decode-uri/DecodeURI';
import { GroupByPipe } from './group-by/GroupBy';
import * as ɵngcc0 from '@angular/core';
export class NovoPipesModule {
}
NovoPipesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoPipesModule });
NovoPipesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoPipesModule_Factory(t) { return new (t || NovoPipesModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoPipesModule, { declarations: function () { return [PluralPipe, DecodeURIPipe, GroupByPipe]; }, exports: function () { return [PluralPipe, DecodeURIPipe, GroupByPipe]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoPipesModule, [{
        type: NgModule,
        args: [{
                declarations: [PluralPipe, DecodeURIPipe, GroupByPipe],
                exports: [PluralPipe, DecodeURIPipe, GroupByPipe]
            }]
    }], null, null); })();
