// novo-elements/esm2015/elements/dropdown/Dropdown.module.js
// NG2
import { NgModule } from '@angular/core';
// APP
import { NovoDropdownElement, NovoItemElement, NovoDropDownItemHeaderElement, NovoDropdownListElement } from './Dropdown';
import { NovoOverlayModule } from '../overlay/Overlay.module';
import * as ɵngcc0 from '@angular/core';
export class NovoDropdownModule {
}
NovoDropdownModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoDropdownModule });
NovoDropdownModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoDropdownModule_Factory(t) { return new (t || NovoDropdownModule)(); }, imports: [[NovoOverlayModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoDropdownModule, { declarations: function () { return [NovoDropdownElement, NovoItemElement, NovoDropdownListElement, NovoDropDownItemHeaderElement]; }, imports: function () { return [NovoOverlayModule]; }, exports: function () { return [NovoDropdownElement, NovoItemElement, NovoDropdownListElement, NovoDropDownItemHeaderElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoDropdownModule, [{
        type: NgModule,
        args: [{
                imports: [NovoOverlayModule],
                declarations: [NovoDropdownElement, NovoItemElement, NovoDropdownListElement, NovoDropDownItemHeaderElement],
                exports: [NovoDropdownElement, NovoItemElement, NovoDropdownListElement, NovoDropDownItemHeaderElement]
            }]
    }], null, null); })();
