// novo-elements/esm2015/elements/search/SearchBox.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoOverlayModule } from '../overlay/Overlay.module';
import { NovoButtonModule } from './../button/Button.module';
import { NovoPickerModule } from './../picker/Picker.module';
import { NovoTooltipModule } from './../tooltip/Tooltip.module';
import { NovoSearchBoxElement } from './SearchBox';
import * as ɵngcc0 from '@angular/core';
export class NovoSearchBoxModule {
}
NovoSearchBoxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoSearchBoxModule });
NovoSearchBoxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoSearchBoxModule_Factory(t) { return new (t || NovoSearchBoxModule)(); }, imports: [[CommonModule, NovoButtonModule, NovoPickerModule, NovoTooltipModule, NovoOverlayModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoSearchBoxModule, { declarations: function () { return [NovoSearchBoxElement]; }, imports: function () { return [CommonModule, NovoButtonModule, NovoPickerModule, NovoTooltipModule, NovoOverlayModule]; }, exports: function () { return [NovoSearchBoxElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoSearchBoxModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NovoButtonModule, NovoPickerModule, NovoTooltipModule, NovoOverlayModule],
                declarations: [NovoSearchBoxElement],
                exports: [NovoSearchBoxElement]
            }]
    }], null, null); })();
