// novo-elements/esm2015/elements/search/SearchBox.js
// NG2
import { Component, Input, Output, EventEmitter, ViewChild, forwardRef, ElementRef, HostBinding, ChangeDetectorRef, NgZone, ChangeDetectionStrategy, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { TAB, ENTER, ESCAPE } from '@angular/cdk/keycodes';
// APP
import { NovoOverlayTemplateComponent } from '../overlay/Overlay';
import { NovoLabelService } from '../../services/novo-label-service';
// Value accessor for the component (supports ngModel)
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../services/novo-label-service';
import * as ɵngcc2 from '../button/Button';
import * as ɵngcc3 from '../tooltip/Tooltip.directive';
import * as ɵngcc4 from '../overlay/Overlay';

const _c0 = ["input"];
const _c1 = ["*"];
const SEARCH_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => NovoSearchBoxElement),
    multi: true,
};
export class NovoSearchBoxElement {
    constructor(element, labels, _changeDetectorRef, _zone) {
        this.element = element;
        this.labels = labels;
        this._changeDetectorRef = _changeDetectorRef;
        this._zone = _zone;
        this.icon = 'search';
        this.placeholder = 'Search...';
        this.alwaysOpen = false;
        this.theme = 'positive';
        this.closeOnSelect = true;
        this.searchChanged = new EventEmitter();
        this.focused = false;
        /** View -> model callback called when value changes */
        this._onChange = () => { };
        /** View -> model callback called when autocomplete has been touched */
        this._onTouched = () => { };
    }
    /**
     * @name showFasterFind
     * @description This function shows the picker and adds the active class (for animation)
     */
    showSearch(event, forceClose = false) {
        if (!this.panelOpen) {
            // Reset search
            // Set focus on search
            setTimeout(() => {
                const element = this.input.nativeElement;
                if (element) {
                    element.focus();
                }
            }, 10);
        }
    }
    onFocus() {
        this._zone.run(() => {
            this.focused = true;
            this.openPanel();
        });
    }
    onBlur() {
        this.focused = false;
    }
    /** BEGIN: Convenient Panel Methods. */
    openPanel() {
        this.overlay.openPanel();
    }
    closePanel() {
        this.overlay.closePanel();
    }
    get panelOpen() {
        return this.overlay && this.overlay.panelOpen;
    }
    get active() {
        return this.panelOpen || this.alwaysOpen;
    }
    /** END: Convenient Panel Methods. */
    _handleKeydown(event) {
        if ((event.keyCode === ESCAPE || event.keyCode === ENTER || event.keyCode === TAB) && this.panelOpen) {
            this.closePanel();
            event.stopPropagation();
        }
    }
    _handleInput(event) {
        if (document.activeElement === event.target) {
            this._onChange(event.target.value);
            if (this.debounceSearchChange) {
                clearTimeout(this.debounceSearchChange);
            }
            this.debounceSearchChange = setTimeout(() => {
                this.searchChanged.emit(event.target.value);
            }, 400);
        }
    }
    writeValue(value) {
        this._setValue(value);
    }
    registerOnChange(fn) {
        this._onChange = fn;
    }
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    _setValue(value) {
        this.value = value;
        let toDisplay = value;
        if (value && this.displayField) {
            toDisplay = value.hasOwnProperty(this.displayField) ? value[this.displayField] : value;
        }
        // Simply falling back to an empty string if the display value is falsy does not work properly.
        // The display value can also be the number zero and shouldn't fall back to an empty string.
        this.displayValue = toDisplay ? toDisplay : '';
        this.input.nativeElement.value = this.displayValue;
        this._changeDetectorRef.markForCheck();
    }
    /**
     * This method closes the panel, and if a value is specified, also sets the associated
     * control to that value. It will also mark the control as dirty if this interaction
     * stemmed from the user.
     */
    setValueAndClose(event) {
        if (event && event.value) {
            this._setValue(event.value);
            this._onChange(event.value);
        }
        this.closePanel();
    }
    /**
     * Clear any previous selected option and emit a selection change event for this option
     */
    clearValue(skip) {
        this.writeValue(null);
        this._onChange(null);
    }
}
NovoSearchBoxElement.ɵfac = function NovoSearchBoxElement_Factory(t) { return new (t || NovoSearchBoxElement)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NovoLabelService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
NovoSearchBoxElement.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NovoSearchBoxElement, selectors: [["novo-search"]], viewQuery: function NovoSearchBoxElement_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(NovoOverlayTemplateComponent, true);
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.overlay = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.input = _t.first);
    } }, hostVars: 4, hostBindings: function NovoSearchBoxElement_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("focused", ctx.focused)("active", ctx.active);
    } }, inputs: { icon: "icon", placeholder: "placeholder", alwaysOpen: "alwaysOpen", theme: "theme", closeOnSelect: "closeOnSelect", displayValue: "displayValue", name: "name", displayField: "displayField", hint: "hint" }, outputs: { searchChanged: "searchChanged" }, features: [ɵngcc0.ɵɵProvidersFeature([SEARCH_VALUE_ACCESSOR])], ngContentSelectors: _c1, decls: 5, vars: 8, consts: [["theme", "fab", "tooltipPosition", "bottom", "data-automation-id", "novo-search-fab", 3, "color", "icon", "tooltip", "click"], ["type", "text", "data-automation-id", "novo-search-input", 3, "focus", "blur", "keydown", "input"], ["input", ""], ["position", "above-below", 3, "parent", "closeOnSelect", "select", "closing"]], template: function NovoSearchBoxElement_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function NovoSearchBoxElement_Template_button_click_0_listener() { return ctx.showSearch(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(1, "input", 1, 2);
        ɵngcc0.ɵɵlistener("focus", function NovoSearchBoxElement_Template_input_focus_1_listener() { return ctx.onFocus(); })("blur", function NovoSearchBoxElement_Template_input_blur_1_listener() { return ctx.onBlur(); })("keydown", function NovoSearchBoxElement_Template_input_keydown_1_listener($event) { return ctx._handleKeydown($event); })("input", function NovoSearchBoxElement_Template_input_input_1_listener($event) { return ctx._handleInput($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "novo-overlay-template", 3);
        ɵngcc0.ɵɵlistener("select", function NovoSearchBoxElement_Template_novo_overlay_template_select_3_listener() { return ctx.closePanel(); })("closing", function NovoSearchBoxElement_Template_novo_overlay_template_closing_3_listener() { return ctx.onBlur(); });
        ɵngcc0.ɵɵprojection(4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("color", ctx.theme)("icon", ctx.icon)("tooltip", ctx.hint);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("name", ctx.name)("value", ctx.displayValue)("placeholder", ctx.placeholder);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("parent", ctx.element)("closeOnSelect", ctx.closeOnSelect);
    } }, directives: [ɵngcc2.NovoButtonElement, ɵngcc3.TooltipDirective, ɵngcc4.NovoOverlayTemplateComponent], encapsulation: 2, changeDetection: 0 });
NovoSearchBoxElement.ctorParameters = () => [
    { type: ElementRef },
    { type: NovoLabelService },
    { type: ChangeDetectorRef },
    { type: NgZone }
];
NovoSearchBoxElement.propDecorators = {
    name: [{ type: Input }],
    icon: [{ type: Input }],
    placeholder: [{ type: Input }],
    alwaysOpen: [{ type: Input }],
    theme: [{ type: Input }],
    closeOnSelect: [{ type: Input }],
    displayField: [{ type: Input }],
    displayValue: [{ type: Input }],
    hint: [{ type: Input }],
    searchChanged: [{ type: Output }],
    focused: [{ type: HostBinding, args: ['class.focused',] }],
    overlay: [{ type: ViewChild, args: [NovoOverlayTemplateComponent,] }],
    input: [{ type: ViewChild, args: ['input', { static: true },] }],
    active: [{ type: HostBinding, args: ['class.active',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoSearchBoxElement, [{
        type: Component,
        args: [{
                selector: 'novo-search',
                providers: [SEARCH_VALUE_ACCESSOR],
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <!-- SEARCH ICON -->
    <button
      theme="fab"
      [color]="theme"
      [icon]="icon"
      (click)="showSearch()"
      [tooltip]="hint"
      tooltipPosition="bottom"
      data-automation-id="novo-search-fab"
    ></button>
    <!-- SEARCH INPUT -->
    <input
      type="text"
      [attr.name]="name"
      [attr.value]="displayValue"
      [attr.placeholder]="placeholder"
      (focus)="onFocus()"
      (blur)="onBlur()"
      (keydown)="_handleKeydown($event)"
      (input)="_handleInput($event)"
      #input
      data-automation-id="novo-search-input"
    />
    <!-- SEARCH OVERLAY -->
    <novo-overlay-template
      [parent]="element"
      [closeOnSelect]="closeOnSelect"
      position="above-below"
      (select)="closePanel()"
      (closing)="onBlur()"
    >
      <ng-content></ng-content>
    </novo-overlay-template>
  `
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.NovoLabelService }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.NgZone }]; }, { icon: [{
            type: Input
        }], placeholder: [{
            type: Input
        }], alwaysOpen: [{
            type: Input
        }], theme: [{
            type: Input
        }], closeOnSelect: [{
            type: Input
        }], searchChanged: [{
            type: Output
        }], focused: [{
            type: HostBinding,
            args: ['class.focused']
        }], active: [{
            type: HostBinding,
            args: ['class.active']
        }], displayValue: [{
            type: Input
        }], name: [{
            type: Input
        }], displayField: [{
            type: Input
        }], hint: [{
            type: Input
        }], overlay: [{
            type: ViewChild,
            args: [NovoOverlayTemplateComponent]
        }], input: [{
            type: ViewChild,
            args: ['input', { static: true }]
        }] }); })();
