// novo-elements/esm2015/elements/value/Value.module.js
// NG2
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
// APP
import { EntityList } from './EntityList';
import { RenderPipe } from './Render';
import { NovoValueElement } from './Value';
import * as ɵngcc0 from '@angular/core';
export class NovoValueModule {
}
NovoValueModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoValueModule });
NovoValueModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoValueModule_Factory(t) { return new (t || NovoValueModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoValueModule, { declarations: function () { return [NovoValueElement, RenderPipe, EntityList]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NovoValueElement, RenderPipe, EntityList]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoValueModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NovoValueElement, RenderPipe, EntityList],
                exports: [NovoValueElement, RenderPipe, EntityList]
            }]
    }], null, null); })();
