// novo-elements/esm2015/services/template/NovoTemplateService.js
// NG2
import { Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class NovoTemplateService {
    constructor() {
        this.templates = {
            default: {},
            custom: {},
        };
    }
    getAll() {
        const templates = {};
        const customTemplateTypes = Object.keys(this.templates.custom);
        const defaultTemplateTypes = Object.keys(this.templates.default);
        defaultTemplateTypes.forEach((type) => {
            templates[type] = this.templates.default[type];
        });
        customTemplateTypes.forEach((type) => {
            templates[type] = this.templates.custom[type];
        });
        return templates;
    }
    addDefault(key, template) {
        this.templates.default[key] = template;
    }
    addCustom(key, template) {
        this.templates.custom[key] = template;
    }
}
NovoTemplateService.ɵfac = function NovoTemplateService_Factory(t) { return new (t || NovoTemplateService)(); };
NovoTemplateService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NovoTemplateService, factory: NovoTemplateService.ɵfac });
NovoTemplateService.ctorParameters = () => [];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoTemplateService, [{
        type: Injectable
    }], function () { return []; }, null); })();
