// novo-elements/esm2015/elements/common/novo-template/novo-template.directive.js
import { Directive, Input, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class NovoTemplate {
    constructor(template) {
        this.template = template;
    }
    getType() {
        return this.name;
    }
}
NovoTemplate.ɵfac = function NovoTemplate_Factory(t) { return new (t || NovoTemplate)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
NovoTemplate.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NovoTemplate, selectors: [["", "novoTemplate", ""]], inputs: { type: "type", name: ["novoTemplate", "name"] } });
NovoTemplate.ctorParameters = () => [
    { type: TemplateRef }
];
NovoTemplate.propDecorators = {
    type: [{ type: Input }],
    name: [{ type: Input, args: ['novoTemplate',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoTemplate, [{
        type: Directive,
        args: [{
                selector: '[novoTemplate]'
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { type: [{
            type: Input
        }], name: [{
            type: Input,
            args: ['novoTemplate']
        }] }); })();
