// novo-elements/esm2015/elements/button/Button.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoButtonElement } from './Button';
import * as ɵngcc0 from '@angular/core';
export class NovoButtonModule {
}
NovoButtonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoButtonModule });
NovoButtonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoButtonModule_Factory(t) { return new (t || NovoButtonModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoButtonModule, { declarations: function () { return [NovoButtonElement]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NovoButtonElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoButtonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NovoButtonElement],
                exports: [NovoButtonElement]
            }]
    }], null, null); })();
