// novo-elements/esm2015/elements/multi-picker/MultiPicker.module.js
// NG2
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
// APP
import { NovoPickerModule } from './../picker/Picker.module';
import { NovoChipsModule } from './../chips/Chips.module';
import { NovoMultiPickerElement } from './MultiPicker';
import * as ɵngcc0 from '@angular/core';
export class NovoMultiPickerModule {
}
NovoMultiPickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoMultiPickerModule });
NovoMultiPickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoMultiPickerModule_Factory(t) { return new (t || NovoMultiPickerModule)(); }, imports: [[CommonModule, FormsModule, NovoPickerModule, NovoChipsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoMultiPickerModule, { declarations: function () { return [NovoMultiPickerElement]; }, imports: function () { return [CommonModule, FormsModule, NovoPickerModule, NovoChipsModule]; }, exports: function () { return [NovoMultiPickerElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoMultiPickerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, NovoPickerModule, NovoChipsModule],
                declarations: [NovoMultiPickerElement],
                exports: [NovoMultiPickerElement]
            }]
    }], null, null); })();
