// novo-elements/esm2015/elements/picker/extras/picker-results/PickerResults.js
// NG2
import { Component, ElementRef, ChangeDetectorRef } from '@angular/core';
// APP
import { NovoLabelService } from '../../../../services/novo-label-service';
import { BasePickerResults } from '../base-picker-results/BasePickerResults';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../../services/novo-label-service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '../../../list/List';
import * as ɵngcc4 from '../../../loading/Loading';

function PickerResults_novo_list_0_novo_list_item_1_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "novo-list-item", 7);
    ɵngcc0.ɵɵlistener("click", function PickerResults_novo_list_0_novo_list_item_1_Template_novo_list_item_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.selectMatch($event); })("mouseenter", function PickerResults_novo_list_0_novo_list_item_1_Template_novo_list_item_mouseenter_0_listener() { ɵngcc0.ɵɵrestoreView(_r8); const match_r6 = ctx.$implicit; const ctx_r9 = ɵngcc0.ɵɵnextContext(2); return ctx_r9.selectActive(match_r6); });
    ɵngcc0.ɵɵelementStart(1, "item-content");
    ɵngcc0.ɵɵelement(2, "span", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const match_r6 = ctx.$implicit;
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("active", match_r6 === ctx_r4.activeMatch)("disabled", ctx_r4.preselected(match_r6));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("innerHtml", ctx_r4.highlight(match_r6.label, ctx_r4.term), ɵngcc0.ɵɵsanitizeHtml);
} }
function PickerResults_novo_list_0_novo_loading_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "novo-loading", 9);
} }
function PickerResults_novo_list_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "novo-list", 4);
    ɵngcc0.ɵɵtemplate(1, PickerResults_novo_list_0_novo_list_item_1_Template, 3, 5, "novo-list-item", 5);
    ɵngcc0.ɵɵtemplate(2, PickerResults_novo_list_0_novo_loading_2_Template, 1, 0, "novo-loading", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.matches);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isLoading && ctx_r0.matches.length > 0);
} }
function PickerResults_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵelement(1, "novo-loading", 9);
    ɵngcc0.ɵɵelementEnd();
} }
function PickerResults_p_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 11);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.labels.pickerError);
} }
function PickerResults_p_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.getEmptyMessage());
} }
export class PickerResults extends BasePickerResults {
    constructor(element, labels, ref) {
        super(element, ref);
        this.labels = labels;
    }
    get hasNonErrorMessage() {
        return !this.isLoading && !this.matches.length && !this.hasError;
    }
    getEmptyMessage() {
        if (this.shouldShowMessageForZeroLengthSearch()) {
            // this property comes from Field Interactions
            return this.config.emptyPickerMessage;
        }
        else {
            return this.term === '' ? this.labels.pickerTextFieldEmpty : this.labels.pickerEmpty;
        }
    }
    shouldShowMessageForZeroLengthSearch() {
        return this.config && this.config.minSearchLength === 0 && this.term === '' && this.config.emptyPickerMessage;
    }
    getListElement() {
        return this.element.nativeElement.querySelector('novo-list');
    }
}
PickerResults.ɵfac = function PickerResults_Factory(t) { return new (t || PickerResults)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NovoLabelService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
PickerResults.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: PickerResults, selectors: [["picker-results"]], hostAttrs: [1, "active"], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 4, vars: 4, consts: [["direction", "vertical", 4, "ngIf"], ["class", "picker-loader", 4, "ngIf"], ["class", "picker-error", 4, "ngIf"], ["class", "picker-null-results", 4, "ngIf"], ["direction", "vertical"], [3, "active", "disabled", "click", "mouseenter", 4, "ngFor", "ngForOf"], ["theme", "line", 4, "ngIf"], [3, "click", "mouseenter"], [3, "innerHtml"], ["theme", "line"], [1, "picker-loader"], [1, "picker-error"], [1, "picker-null-results"]], template: function PickerResults_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, PickerResults_novo_list_0_Template, 3, 2, "novo-list", 0);
        ɵngcc0.ɵɵtemplate(1, PickerResults_div_1_Template, 2, 0, "div", 1);
        ɵngcc0.ɵɵtemplate(2, PickerResults_p_2_Template, 2, 1, "p", 2);
        ɵngcc0.ɵɵtemplate(3, PickerResults_p_3_Template, 2, 1, "p", 3);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.matches.length > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isLoading && ctx.matches.length === 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasError);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasNonErrorMessage);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.NovoListElement, ɵngcc2.NgForOf, ɵngcc3.NovoListItemElement, ɵngcc3.NovoItemContentElement, ɵngcc4.NovoLoadingElement], encapsulation: 2 });
PickerResults.ctorParameters = () => [
    { type: ElementRef },
    { type: NovoLabelService },
    { type: ChangeDetectorRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PickerResults, [{
        type: Component,
        args: [{
                selector: 'picker-results',
                host: {
                    class: 'active'
                },
                template: `
    <novo-list *ngIf="matches.length > 0" direction="vertical">
      <novo-list-item
        *ngFor="let match of matches"
        (click)="selectMatch($event)"
        [class.active]="match === activeMatch"
        (mouseenter)="selectActive(match)"
        [class.disabled]="preselected(match)"
      >
        <item-content> <span [innerHtml]="highlight(match.label, term)"></span> </item-content>
      </novo-list-item>
      <novo-loading *ngIf="isLoading && matches.length > 0" theme="line"></novo-loading>
    </novo-list>
    <div class="picker-loader" *ngIf="isLoading && matches.length === 0"><novo-loading theme="line"></novo-loading></div>
    <p class="picker-error" *ngIf="hasError">{{ labels.pickerError }}</p>
    <p class="picker-null-results" *ngIf="hasNonErrorMessage">{{ getEmptyMessage() }}</p>
  `
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.NovoLabelService }, { type: ɵngcc0.ChangeDetectorRef }]; }, null); })();
