// novo-elements/esm2015/elements/common/common.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NovoTemplate } from './novo-template/novo-template.directive';
import * as ɵngcc0 from '@angular/core';
export class NovoCommonModule {
}
NovoCommonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoCommonModule });
NovoCommonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoCommonModule_Factory(t) { return new (t || NovoCommonModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoCommonModule, { declarations: function () { return [NovoTemplate]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NovoTemplate]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoCommonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [NovoTemplate],
                declarations: [NovoTemplate]
            }]
    }], null, null); })();
