// novo-elements/esm2015/elements/popover/PopOver.module.js
// NG2
import { NgModule } from '@angular/core';
// APP
import { PopOverContent } from './PopOverContent';
import { PopOverDirective } from './PopOver';
import * as ɵngcc0 from '@angular/core';
export class NovoPopOverModule {
}
NovoPopOverModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoPopOverModule });
NovoPopOverModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoPopOverModule_Factory(t) { return new (t || NovoPopOverModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoPopOverModule, { declarations: function () { return [PopOverContent, PopOverDirective]; }, exports: function () { return [PopOverContent, PopOverDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoPopOverModule, [{
        type: NgModule,
        args: [{
                declarations: [PopOverContent, PopOverDirective],
                exports: [PopOverContent, PopOverDirective]
            }]
    }], null, null); })();
