// novo-elements/esm2015/elements/time-picker/TimePicker.module.js
// NG2
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
// Vendor
import { TextMaskModule } from 'angular2-text-mask';
// APP
import { NovoOverlayModule } from '../overlay/Overlay.module';
import { NovoListModule } from '../list/List.module';
import { NovoTimePickerElement } from './TimePicker';
import { NovoTimePickerInputElement } from './TimePickerInput';
import * as ɵngcc0 from '@angular/core';
export class NovoTimePickerModule {
}
NovoTimePickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoTimePickerModule });
NovoTimePickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoTimePickerModule_Factory(t) { return new (t || NovoTimePickerModule)(); }, imports: [[CommonModule, FormsModule, TextMaskModule, NovoOverlayModule, NovoListModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoTimePickerModule, { declarations: function () { return [NovoTimePickerElement, NovoTimePickerInputElement]; }, imports: function () { return [CommonModule, FormsModule, TextMaskModule, NovoOverlayModule, NovoListModule]; }, exports: function () { return [NovoTimePickerElement, NovoTimePickerInputElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoTimePickerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, TextMaskModule, NovoOverlayModule, NovoListModule],
                declarations: [NovoTimePickerElement, NovoTimePickerInputElement],
                exports: [NovoTimePickerElement, NovoTimePickerInputElement]
            }]
    }], null, null); })();
