// novo-elements/esm2015/elements/ckeditor/CKEditor.module.js
// NG2
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
// APP
import { NovoCKEditorElement } from './CKEditor';
import * as ɵngcc0 from '@angular/core';
export class NovoNovoCKEditorModule {
}
NovoNovoCKEditorModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoNovoCKEditorModule });
NovoNovoCKEditorModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoNovoCKEditorModule_Factory(t) { return new (t || NovoNovoCKEditorModule)(); }, imports: [[CommonModule, FormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoNovoCKEditorModule, { declarations: function () { return [NovoCKEditorElement]; }, imports: function () { return [CommonModule, FormsModule]; }, exports: function () { return [NovoCKEditorElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoNovoCKEditorModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule],
                declarations: [NovoCKEditorElement],
                exports: [NovoCKEditorElement]
            }]
    }], null, null); })();
