// novo-elements/esm2015/elements/date-picker/DatePicker.module.js
// NG2
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
// Vendor
import { TextMaskModule } from 'angular2-text-mask';
// APP
import { NovoOverlayModule } from '../overlay/Overlay.module';
import { NovoDatePickerElement } from './DatePicker';
import { NovoDatePickerInputElement } from './DatePickerInput';
import * as ɵngcc0 from '@angular/core';
export class NovoDatePickerModule {
}
NovoDatePickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoDatePickerModule });
NovoDatePickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoDatePickerModule_Factory(t) { return new (t || NovoDatePickerModule)(); }, imports: [[CommonModule, FormsModule, NovoOverlayModule, TextMaskModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoDatePickerModule, { declarations: function () { return [NovoDatePickerElement, NovoDatePickerInputElement]; }, imports: function () { return [CommonModule, FormsModule, NovoOverlayModule, TextMaskModule]; }, exports: function () { return [NovoDatePickerElement, NovoDatePickerInputElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoDatePickerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, NovoOverlayModule, TextMaskModule],
                declarations: [NovoDatePickerElement, NovoDatePickerInputElement],
                exports: [NovoDatePickerElement, NovoDatePickerInputElement]
            }]
    }], null, null); })();
