// novo-elements/esm2015/elements/quick-note/QuickNote.module.js
// NG2
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
// APP
import { NovoLoadingModule } from './../loading/Loading.module';
import { NovoListModule } from './../list/List.module';
import { QuickNoteElement } from './QuickNote';
import { QuickNoteResults } from './extras/quick-note-results/QuickNoteResults';
import * as ɵngcc0 from '@angular/core';
export class NovoQuickNoteModule {
}
NovoQuickNoteModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoQuickNoteModule });
NovoQuickNoteModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoQuickNoteModule_Factory(t) { return new (t || NovoQuickNoteModule)(); }, imports: [[CommonModule, FormsModule, NovoLoadingModule, NovoListModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoQuickNoteModule, { declarations: function () { return [QuickNoteElement, QuickNoteResults]; }, imports: function () { return [CommonModule, FormsModule, NovoLoadingModule, NovoListModule]; }, exports: function () { return [QuickNoteElement, QuickNoteResults]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoQuickNoteModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, NovoLoadingModule, NovoListModule],
                declarations: [QuickNoteElement, QuickNoteResults],
                exports: [QuickNoteElement, QuickNoteResults]
            }]
    }], null, null); })();
