// novo-elements/esm2015/elements/places/places.module.js
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NovoListModule } from '../list/List.module';
import { PlacesListComponent } from './places.component';
import { GooglePlacesService } from './places.service';
import * as ɵngcc0 from '@angular/core';
export class GooglePlacesModule {
}
GooglePlacesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: GooglePlacesModule });
GooglePlacesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function GooglePlacesModule_Factory(t) { return new (t || GooglePlacesModule)(); }, providers: [GooglePlacesService], imports: [[CommonModule, HttpClientModule, FormsModule, NovoListModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(GooglePlacesModule, { declarations: function () { return [PlacesListComponent]; }, imports: function () { return [CommonModule, HttpClientModule, FormsModule, NovoListModule]; }, exports: function () { return [PlacesListComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GooglePlacesModule, [{
        type: NgModule,
        args: [{
                declarations: [PlacesListComponent],
                imports: [CommonModule, HttpClientModule, FormsModule, NovoListModule],
                exports: [PlacesListComponent],
                providers: [GooglePlacesService]
            }]
    }], null, null); })();
