// novo-elements/esm2015/elements/toast/Toast.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoToastElement } from './Toast';
import * as ɵngcc0 from '@angular/core';
export class NovoToastModule {
}
NovoToastModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoToastModule });
NovoToastModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoToastModule_Factory(t) { return new (t || NovoToastModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoToastModule, { declarations: function () { return [NovoToastElement]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NovoToastElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoToastModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NovoToastElement],
                exports: [NovoToastElement]
            }]
    }], null, null); })();
