// novo-elements/esm2015/elements/overlay/Overlay.module.js
// NG2
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { ScrollingModule } from '@angular/cdk/scrolling';
// APP
import { NovoOverlayTemplateComponent } from './Overlay';
import * as ɵngcc0 from '@angular/core';
export class NovoOverlayModule {
}
NovoOverlayModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoOverlayModule });
NovoOverlayModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoOverlayModule_Factory(t) { return new (t || NovoOverlayModule)(); }, imports: [[CommonModule, FormsModule, OverlayModule, ScrollingModule], ScrollingModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoOverlayModule, { declarations: function () { return [NovoOverlayTemplateComponent]; }, imports: function () { return [CommonModule, FormsModule, OverlayModule, ScrollingModule]; }, exports: function () { return [NovoOverlayTemplateComponent, ScrollingModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoOverlayModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, OverlayModule, ScrollingModule],
                declarations: [NovoOverlayTemplateComponent],
                exports: [NovoOverlayTemplateComponent, ScrollingModule]
            }]
    }], null, null); })();
