// novo-elements/esm2015/elements/modal/Modal.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoButtonModule } from './../button/Button.module';
import { NovoModalContainerElement, NovoModalElement, NovoModalNotificationElement } from './Modal';
import * as ɵngcc0 from '@angular/core';
export class NovoModalModule {
}
NovoModalModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoModalModule });
NovoModalModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoModalModule_Factory(t) { return new (t || NovoModalModule)(); }, imports: [[CommonModule, NovoButtonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoModalModule, { declarations: function () { return [NovoModalContainerElement, NovoModalElement, NovoModalNotificationElement]; }, imports: function () { return [CommonModule, NovoButtonModule]; }, exports: function () { return [NovoModalElement, NovoModalNotificationElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoModalModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NovoButtonModule],
                declarations: [NovoModalContainerElement, NovoModalElement, NovoModalNotificationElement],
                exports: [NovoModalElement, NovoModalNotificationElement]
            }]
    }], null, null); })();
