// novo-elements/esm2015/elements/modal/ModalService.js
// NG2
import { Injectable } from '@angular/core';
// APP
import { NovoModalRef, NovoModalParams, NovoModalContainerElement } from './Modal';
import { ComponentUtils } from './../../utils/component-utils/ComponentUtils';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './../../utils/component-utils/ComponentUtils';
export class NovoModalService {
    constructor(componentUtils) {
        this.componentUtils = componentUtils;
    }
    set parentViewContainer(view) {
        this._parentViewContainer = view;
    }
    open(component, scope = {}) {
        if (!this._parentViewContainer) {
            throw new Error('No parent view container specified for the ModalService. Set it inside your main application. \nthis.modalService.parentViewContainer = view (ViewContainerRef)');
        }
        const modal = new NovoModalRef();
        modal.component = component;
        modal.open();
        const providers = [{ provide: NovoModalRef, useValue: modal }, { provide: NovoModalParams, useValue: scope }];
        modal.containerRef = this.componentUtils.append(NovoModalContainerElement, this._parentViewContainer, providers);
        return modal;
    }
}
NovoModalService.ɵfac = function NovoModalService_Factory(t) { return new (t || NovoModalService)(ɵngcc0.ɵɵinject(ɵngcc1.ComponentUtils)); };
NovoModalService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NovoModalService, factory: NovoModalService.ɵfac });
NovoModalService.ctorParameters = () => [
    { type: ComponentUtils }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoModalService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.ComponentUtils }]; }, null); })();
