// novo-elements/esm2015/elements/switch/Switch.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
// APP
import { NovoSwitchElement } from './Switch';
import * as ɵngcc0 from '@angular/core';
export class NovoSwitchModule {
}
NovoSwitchModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoSwitchModule });
NovoSwitchModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoSwitchModule_Factory(t) { return new (t || NovoSwitchModule)(); }, imports: [[CommonModule, FormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoSwitchModule, { declarations: function () { return [NovoSwitchElement]; }, imports: function () { return [CommonModule, FormsModule]; }, exports: function () { return [NovoSwitchElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoSwitchModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule],
                declarations: [NovoSwitchElement],
                exports: [NovoSwitchElement]
            }]
    }], null, null); })();
