// novo-elements/esm2015/elements/slider/Slider.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoButtonModule } from './../button/Button.module';
import { NovoSliderElement } from './Slider';
import * as ɵngcc0 from '@angular/core';
export class NovoSliderModule {
}
NovoSliderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoSliderModule });
NovoSliderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoSliderModule_Factory(t) { return new (t || NovoSliderModule)(); }, imports: [[CommonModule, NovoButtonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoSliderModule, { declarations: function () { return [NovoSliderElement]; }, imports: function () { return [CommonModule, NovoButtonModule]; }, exports: function () { return [NovoSliderElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoSliderModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NovoButtonModule],
                declarations: [NovoSliderElement],
                exports: [NovoSliderElement]
            }]
    }], null, null); })();
