// novo-elements/esm2015/elements/tiles/Tiles.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
// APP
import { NovoTilesElement } from './Tiles';
import * as ɵngcc0 from '@angular/core';
export class NovoTilesModule {
}
NovoTilesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoTilesModule });
NovoTilesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoTilesModule_Factory(t) { return new (t || NovoTilesModule)(); }, imports: [[CommonModule, ReactiveFormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoTilesModule, { declarations: function () { return [NovoTilesElement]; }, imports: function () { return [CommonModule, ReactiveFormsModule]; }, exports: function () { return [NovoTilesElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoTilesModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ReactiveFormsModule],
                declarations: [NovoTilesElement],
                exports: [NovoTilesElement]
            }]
    }], null, null); })();
