// novo-elements/esm2015/elements/category-dropdown/CategoryDropdown.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoCategoryDropdownElement } from './CategoryDropdown';
import { NovoTabModule } from './../tabs/Tabs.module';
import { NovoListModule } from './../list/List.module';
import * as ɵngcc0 from '@angular/core';
export class NovoCategoryDropdownModule {
}
NovoCategoryDropdownModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoCategoryDropdownModule });
NovoCategoryDropdownModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoCategoryDropdownModule_Factory(t) { return new (t || NovoCategoryDropdownModule)(); }, imports: [[CommonModule, NovoTabModule, NovoListModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoCategoryDropdownModule, { declarations: function () { return [NovoCategoryDropdownElement]; }, imports: function () { return [CommonModule, NovoTabModule, NovoListModule]; }, exports: function () { return [NovoCategoryDropdownElement]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoCategoryDropdownModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NovoTabModule, NovoListModule],
                declarations: [NovoCategoryDropdownElement],
                exports: [NovoCategoryDropdownElement]
            }]
    }], null, null); })();
