// novo-elements/esm2015/elements/radio/Radio.module.js
// NG2
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
// APP
import { NovoButtonModule } from '../button/Button.module';
import { NovoRadioElement, NovoRadioGroup } from './Radio';
import * as ɵngcc0 from '@angular/core';
export class NovoRadioModule {
}
NovoRadioModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NovoRadioModule });
NovoRadioModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NovoRadioModule_Factory(t) { return new (t || NovoRadioModule)(); }, imports: [[CommonModule, NovoButtonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NovoRadioModule, { declarations: function () { return [NovoRadioElement, NovoRadioGroup]; }, imports: function () { return [CommonModule, NovoButtonModule]; }, exports: function () { return [NovoRadioElement, NovoRadioGroup]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NovoRadioModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NovoButtonModule],
                declarations: [NovoRadioElement, NovoRadioGroup],
                exports: [NovoRadioElement, NovoRadioGroup]
            }]
    }], null, null); })();
